### PR TITLE
feat: update instrumentation

### PR DESCRIPTION
**Pull request by Stebalien**
We're not yet taking advantage of it, but this uses the new
instrumentation logic.

### CODE REVIEW

File: fvm/src/engine.rs [hunk 1]
```diff
@@ -564,18 +555,3 @@ impl Engine {
         store
     }
 }
-
-// Workaround for https://github.com/filecoin-project/ref-fvm/issues/602
-//
-// This removes the out-of-order data count section, if it exists, and re-inserts it (with the
-// correct data-count).
-fn fix_wasm_sections(module: &mut elements::Module) {
```
Comment (Stebalien) [inline fvm/src/engine.rs:572]: We no longer need this hack.

File: fvm/src/engine.rs [hunk 2]
```diff
@@ -392,17 +391,13 @@ impl Engine {
...
-
-        let m = deserialize_buffer(raw_wasm)?;
+        use fvm_wasm_instrument::{gas_metering, stack_limiter};
 
         // stack limiter adds post/pre-ambles to call instructions; We want to do that
         // before injecting gas accounting calls to avoid this overhead in every single
         // block of code.
-        let m =
-            inject_stack_limiter(m, self.0.config.max_wasm_stack).map_err(anyhow::Error::msg)?;
+        let raw_wasm = stack_limiter::inject(raw_wasm, self.0.config.max_wasm_stack)
```
Comment (magik6k) [inline fvm/src/engine.rs:399]: Do we validate the bytecode anywhere before this?

Comment (magik6k) [reply fvm/src/engine.rs:399]: Assuming that this will be done on actor install when wasm contracts are a thing

**Review by magik6k — APPROVED** [review]

File: fvm/src/engine.rs (hunk 2 above)
Comment (Stebalien) [reply fvm/src/engine.rs:399]: We do at the top of this function as well.

**Comment by Stebalien** [comment]
This has reduced gas costs of the SimpleCoin contract (see https://github.com/filecoin-project/fvm-bench/) by 4. That is, 4 measly gas.

So, now we need to know _why_.

**Comment by Stebalien** [comment]
We weren't charging for memory instructions at all. Fixed.

**Pull request merged**